src/aiida_quantumespresso/workflows/xspectra/crystal.py

```diff
@@ -0,0 +1,676 @@
+            if len(elements_not_present) > 0:
+                raise ValueError(
+                    f'The following elements: {elements_not_present} are not present in the'
+                    f' structure ({elements_present}) provided.'
+                    )
+            else:
+                builder.elements_list = orm.List(elements_list)
+                for element in pseudos:
+                    core_hole_pseudos[element] = pseudos[element]['core_hole']
+                    gipaw_pseudos[element] = pseudos[element]['gipaw']
+        # if no elements list is given, we instead initalise the pseudos dict with all
+        # elements in the structure
+        else:
+            for element in pseudos:
+                core_hole_pseudos[element] = pseudos[element]['core_hole']
+                gipaw_pseudos[element] = pseudos[element]['gipaw']
+        builder.core_hole_pseudos = core_hole_pseudos
+        builder.gipaw_pseudos = gipaw_pseudos
+        if core_hole_treatments:
+            builder.core_hole_treatments = orm.Dict(dict=core_hole_treatments)
+        if core_wfc_data:
+            builder.core_wfc_data = core_wfc_data
+        elif upf2plotcore_code:
+            builder.upf2plotcore_code = upf2plotcore_code
+        else:
+            raise ValueError(
+                'No code node for upf2plotcore.sh or core wavefunction data were provided.'
+            )
+        # pylint: enable=no-member
+        return builder
+
+
+    @staticmethod
+    def validate_inputs(inputs, _):
+        """Validate the inputs before launching the WorkChain."""
+        structure = inputs['structure']
+        kinds_present = [kind.name for kind in structure.kinds]
+        elements_present = [kind.symbol for kind in structure.kinds]
+        if 'elements_list' in inputs:
+            absorbing_elements_list = sorted(inputs['elements_list'])
+            extra_elements = []
+            for element in absorbing_elements_list:
+                if element not in elements_present:
+                    extra_elements.append(element)
+            if len(extra_elements) > 0:
+                raise ValidationError(
+                    f'The some elements in ``elements_list`` {extra_elements} do not exist in the'
+                    f' structure provided {elements_present}.'
+                )
+        else:
+            absorbing_elements_list = sorted([kind.symbol for kind in structure.kinds])
+        abs_atom_marker = inputs['abs_atom_marker'].value
+        if abs_atom_marker in kinds_present:
+            raise ValidationError(
+                f'The marker given for the absorbing atom ("{abs_atom_marker}") matches an existing Kind in the '
+                f'input structure ({kinds_present}).'
+            )
+
+        if not inputs['core']['get_powder_spectrum'].value:
+            raise ValidationError(
+                'The ``get_powder_spectrum`` input for the XspectraCoreWorkChain namespace must be ``True``.'
```

Comment:
This is to ensure that the WorkChain returns the expected `powder_spectrum` output - otherwise the `CoreWorkChain` sub-processes won't return it, since `get_powder_spectrum` is `False` by default. Getting the powder spectrum is required, since the intention of the `CrystalWorkChain` is to return the complete powder XAS spectrum for each chosen element in the structure - thus it will need to compute this for each symetrically-inequivalent atom of each element.

To generate a reasonable approximation to the powder spectrum, we compute the XAS along each symetrically-inequivalent periodic vector of the cell and take the average (as detailed in [Brouder *et al* (1990)](dx.doi.org/10.1088/0953-8984/2/3/018)) - thus we need to calculate for 1, 2, or 3 different vectors depending on the spacegroup of the cell.